File: Trees/BinaryTreeJava/src/binarytree/DiagonalTraversal.java
package binarytree;

import java.util.ArrayList;
import java.util.Map;
import java.util.TreeMap;

public class DiagonalTraversal {
    public static void main(String[] args) {
        Node root = null;
        BinaryTree s = new BinaryTree();
        root = s.buildTree(root);
        ArrayList<Integer> ans  = diagonalTraversal(root);
        System.out.println(ans);
        //1 2 4 -1 -1 5 -1 -1 3 6 -1 -1 -1
    }

    public static ArrayList<Integer> diagonalTraversal(Node root){
        ArrayList<Integer> ans = new ArrayList<>();
        Map<Integer , ArrayList<Integer>> map = new TreeMap<>();
        solve(root , map , 0);
        for (Map.Entry<Integer,ArrayList<Integer>> i: map.entrySet()) {
            ans.addAll(i.getValue());
        }
        return ans;
     }

    private static void solve(Node root, Map<Integer, ArrayList<Integer>> map, int hd) {
        if(root == null){
            return ;
        }
        map.putIfAbsent(hd , new ArrayList<>());
        map.get(hd).add(root.data);
        solve(root.left , map , hd+1);
        solve(root.right , map , hd);
    }

}

File: Trees/BinaryTreesC++/diagonalTraversal.cpp
#include <bits/stdc++.h>
using namespace std;

class Node{
	public:
		int data;
		Node* left;
		Node* right;
		Node(int d){
			this->data = d;
			this->left = NULL;
			this->right = NULL;
		}
};
Node* buildTree(Node* root){	
	int data; 
	cin>>data ;
	root = new Node(data);
	if(data == -1){
		return NULL;
	}
	root->left = buildTree(root->left);
	root->right = buildTree(root->right);
	return root;
}

void solve(Node* root , map<int,vector<int>> &map ,int hd ){
	if(root == NULL){
		return ;
	}

	map[hd].push_back(root->data);

	solve(root->left , map , hd+1 );
	solve(root -> right , map , hd);

}

vector<int> diagonalTraversal(Node* root){
	vector<int> ans ;
	map<int,vector<int>> map;
	solve(root , map ,0 );
	for(auto i : map){
		for(auto j : i.second){
			ans.push_back(j);
		}
	}
	return ans;
}

int main(){
	Node* root = NULL ;
	root = buildTree(root);
	vector<int> ans  = diagonalTraversal(root);
	for(auto i : ans){
		cout<< i <<" ";
	}
}
